### ISSUE TITLE
Pivoted document length normalisation [DISCUSSION]

### ISSUE THREAD

**Issue by markroxor** [FreeDiscovery/FreeDiscovery#179]
The repository has implemented [pivoted document length normalisation](http://singhal.info/pivoted-dln.pdf).
But I think that there is a potential bug in [this line](https://github.com/FreeDiscovery/FreeDiscovery/blob/a3763c9/freediscovery/feature_weighting.py#L427) where `norm_alpha` is multiplied with `X_norm` I think that it should be multiplied with `X` instead.

Also I want to have a brief conversation about evaluating the code. Can we get in touch on gitter?

**Comment by rth** [comment]
Thanks for the feedback @markroxor 

> But I think that there is a potential bug in this line where norm_alpha is multiplied with X_norm I think that it should be multiplied with X instead.

Well from the Singhal et al paper,
![untitled](https://user-images.githubusercontent.com/630936/36536571-b3333f2e-17cd-11e8-95ec-9a677cb336b6.png)

in the normalization factor, the slope (`norm_alpha`) is multiplied by the "old normalization" (`X_norm`). There is also a few tests about asymptotic behavior when `norm_alpha=1` [here](https://github.com/FreeDiscovery/FreeDiscovery/blob/55520f5e6a24da0d5e05112809da9ed72c344244/freediscovery/tests/test_feature_weighting.py#L44). 

I could be missing something, but what makes you think there is an issue there? From what I can see in your PR https://github.com/RaRe-Technologies/gensim/pull/1780 the same normalization is used so far.

> Also I want to have a brief conversation about evaluating the code. Can we get in touch on gitter?

Sure, for a general discussion gitter could work, but I think Github would still be better for discussing e.g. if this is a valid issue or not ..

**Comment by rth** [comment]
Also could please rename this issue into something more meaningful (e.g. validity of the pivoted normalization implementation etc).

**Comment by markroxor** [comment]
I might be wrong, but this issue will certainly help in clarifying my doubts.
According to your code - when `norm_alpha = 1`, tfidf will be `tf.idf/X_norm` but shouldn't it be something like `normalize(tf.idf)`?

I might have missed something that I would like to be pointed out.

**Comment by rth** [comment]
> According to your code - when norm_alpha = 1, tfidf will be tf.idf/X_norm but shouldn't it be something like normalize(tf.idf)?

Yes, but `normalize(X)` is `X / row_norms(X)[:, None]` (cf [here](https://github.com/scikit-learn/scikit-learn/blob/a24c8b46/sklearn/preprocessing/data.py#L1438)) and in this implementation, [`X_norm = row_norms(tf.idf)`](https://github.com/FreeDiscovery/FreeDiscovery/blob/a3763c910674eea38099bc83fb98f88e61e6bfc7/freediscovery/feature_weighting.py#L412) . So this should give the same result.  I'm also checking that when `norm_alpha=1` we get the same result as scikit-learn's `CountVectorizer` which does use the regular L2 normalisation function...

> (Also please check gitter for evaluation related queries.)

Could you post them here or open a new issue for them? gitter is blocked by my org firewall at the moment, so I won't be able to check those for some time..

**Comment by rth** [comment]
> scikit-learn's CountVectorizer which does use the regular L2 normalisation function...

No actually it doesn't, it would be a good test to add..

**Comment by markroxor** [comment]
Here's the gitter message - 

I want to do a comparative study on how the pivoted TFIDF is better than the conventional one.
I  tried using TREC as described in the paper but unfortunately it is not freely available.
How can I proceed with this comparative study? The author suggest that the pivot point is the intersection of the relevance and retrieval curve but for real life datasets finding the pivot point is hard following the author's suggested approach.

I have done a very brief comparison in the In [19]: (last third) block of the [ipython notebook](https://render.githubusercontent.com/view/ipynb?commit=1ee449d54047e8c97fc4398552bfd3496c4c95e8&enc_url=68747470733a2f2f7261772e67697468756275736572636f6e74656e742e636f6d2f6d61726b726f786f722f67656e73696d2f316565343439643534303437653863393766633433393835353262666433343936633463393565382f646f63732f6e6f7465626f6f6b732f736b6c6561726e5f6170692e6970796e62&nwo=markroxor%2Fgensim&path=docs%2Fnotebooks%2Fsklearn_api.ipynb&repository_id=69356880&repository_type=Repository#Using-pivoted-document-length-normalisation-with-appropriate-hyperparameters)
but that doesnt verify how this approach is distinct for long-short documents.
Your help is really appreciated.

**Comment by rth** [comment]
> I want to do a comparative study on how the pivoted TFIDF is better than the conventional one.

Well so far we have just tested this empirically with binary classification on a subset of 20 newsgroup dataset in http://freediscovery.io/doc/stable/python/examples/optimize_tfidf.html When looking for the SMART IR weighting scheme that optimizes accuracy, the pivoted TF-IDF does get selected. 

> How can I proceed with this comparative study? The author suggest that the pivot point is the intersection of the relevance and retrieval curve but for real life datasets finding the pivot point is hard following the author's suggested approach.

That's a good question. Maybe @dagr1234 would have some suggestions..

**Comment by markroxor** [comment]
In the ipython notebook I talked about - coincidentally I've also benchmarked on the twenty news group dataset but my aim is to verify the accuracy improvement for long documents in comparison to short documents. As suggested in the paper
> Cosine normalization favors retrieval of short documents.
And the author' s scheme is sort of length independent. How can I show this? Any ideas?

**Comment by rth** [comment]
>> Cosine normalization favors retrieval of short documents.
>
> And the author' s scheme is sort of length independent. How can I show this? Any ideas? 

Well, I imagine with a labeled corpus, you could compute the distribution of length for positive documents. That would be your reference. Then compute the scores for each document in the corpus for either one (or multiple) search queries or a binary classifier. Take the top k documents in the corpus (k to be determined) with the highest scores, and visualize how the length distribution of this subset of the corpus is different from the reference distribution. There should be an effect of favoring short documents with cosine normalization when compared to pivoted normalization.

**Comment by markroxor** [comment]
This approach seems very plausible. Can you fan out some details please -
If I again consider the 20 news group dataset for evaluation and perform binary classification(using only two topics from the dataset).
> you could compute the distribution of length for positive documents.

What do you mean by "positive" documents?
> Then compute the scores for each document in the corpus for either one (or multiple) search queries or a binary classifier

I am planning to compute scores using a pipeline of "tfidf", "LDA", "logistic regression". How to calculate the "scores for each document"? I will get the score of the overall model as shown [here](https://render.githubusercontent.com/view/ipynb?commit=1ee449d54047e8c97fc4398552bfd3496c4c95e8&enc_url=68747470733a2f2f7261772e67697468756275736572636f6e74656e742e636f6d2f6d61726b726f786f722f67656e73696d2f316565343439643534303437653863393766633433393835353262666433343936633463393565382f646f63732f6e6f7465626f6f6b732f736b6c6561726e5f6170692e6970796e62&nwo=markroxor%2Fgensim&path=docs%2Fnotebooks%2Fsklearn_api.ipynb&repository_id=69356880&repository_type=Repository#Using-pivoted-document-length-normalisation-with-appropriate-hyperparameters) which is proportional to its classification accuracy.

It will be clearer if you could use the 20-newsgroup dataset to explain your strategy.

**Comment by dagr1234** [comment]
I really don't have any great ideas for finding a pivot point without ground truth.  I know the TREC data is not publicly available but its fairly easy to sign up for a TREC track and participate in a task and this type of work could easily be done on a TREC track.  I'd strongly encourage that as a good way of doing further tests on this - and I think it'd be something that might yield some interesting results.

**Comment by markroxor** [comment]
I will sign up for TREC. Let's forget about finding the pivot point for the time being.

> Well, I imagine with a labeled corpus, you could compute the distribution of length for positive documents. That would be your reference. Then compute the scores for each document in the corpus for either one (or multiple) search queries or a binary classifier. Take the top k documents in the corpus (k to be determined) with the highest scores, and visualize how the length distribution of this subset of the corpus is different from the reference distribution. There should be an effect of favoring short documents with cosine normalization when compared to pivoted normalization.

I think the above mentioned approach can also be used. I just need little more details, please answer my queries in my last comment.

**Comment by rth** [comment]
>> you could compute the distribution of length for positive documents.
> What do you mean by "positive" documents?

Well, you if you do binary classification documents -- relevant documents.

> I am planning to compute scores using a pipeline of "tfidf", "LDA", "logistic regression". How to calculate the "scores for each document"? I will get the score of the overall model as shown here which is proportional to its classification accuracy.

Not sure what LDA will add to this particular evaluation, it'll also be harder to explain the results. For logistic regression, this could be the scores returned by `LogisticRegression.predict_proba` or `decision_function` methods for each document.

**Comment by markroxor** [comment]
@rth I didn't get my hands on TREC data but I managed with the twenty newsgroup dataset.
I have created a python notebook and a blog comparing both the approaches, I will post the links once they are finished, for you to go through.
Thanks for the discussion.

**Comment by markroxor** [comment]
While trying to find the optimum value of `norm_pivot` I noticed that we get better accuracies as the value of pivot is decreased (mostly to `1`).

Here I have tried to show this using your example with minimal changes of my own.
https://gist.github.com/markroxor/67fed9dfde17a5db25f96fc5971886fb

Can you please explain why this happens?
I have noticed this on other datasets as well.

**Comment by markroxor** [comment]
ping @rth
